### DISCUSSION TITLE
Can logging inherit metadata from spans?

### DISCUSSION THREAD

**Discussion by nickb937** [tokio-rs/tracing#3134] [Q&A]
Most of my functions are decorated with `#[tracing:instrument]` and sometimes I store the span field information in the definition too, or use `span.record()` to store the metadata later calculated in the function.

What I'd like to know is whether it is possible for `info!()` logs to inherit the metadata of the span so that it can be written to stdout as part of the normal log message?

The purpose is that whilst I might look at traces in something like Tempo where that metadata exists, I also like to be able to query the logs system directly. 

Currently I have to duplicate the metadata for each `info!()` statement, which can mean up-to 10 pieces of duplicated metadata written per statement for a given function.

**Comment by mladedav** [comment]
It depends on the logging layer, if you're using fmt subscriber provided in tracing-subscriber, you're probably looking for [with_current_span](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/struct.Layer.html#method.with_current_span) or [with_span_list](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/struct.Layer.html#method.with_span_list).

**Reply by nickb937** [reply]
It's not really the span that I want, as I already have `with_current_span(true)` defined and I'm using `tracing_subscriber::fmt`, it's more that I want all the key-value data associated to the span to be automatically included in the log output i.e.

If i have something like this
```rust
    #[instrument(
        skip_all,
        fields(
            new_policy_plan,
            insurer_id,
            claim_consultant_id,
            appointment_id,
            new_active_contract
        )
    )]
async fn foo() {
/// ...

        let cur_span = tracing::Span::current();
        cur_span.record("policy_plan_id", new_policy_plan.policy_plan_id.as_value());
        cur_span.record("insurer_id", insurer_id.as_value());
        cur_span.record("contract_id", new_active_contract.contract_id.as_value());
        cur_span.record("appointment_id", appointment_id.as_value());
```

Then I would like to be able to write this, to implicitly log those span key-value pairs alongside the message:

```rust
info!("some message")
```

rather than having to explicitly specify them like this for each message:

```rust

        info!(
            policy_plan_id = insurance_attribute.policy_plan_id.as_value(),
            insurer_id = insurer_id.as_value(),
            contract_id = new_active_contract.contract_id.as_value(),
            appointment_id = appointment_id.as_value(),
            "some message"
        );
```

**Reply by mladedav** [reply]
Well, `insurer_id` and `appointment_id` should be present in the output in span fields.

The others wouldn't, because you must specify all span fields beforehand. The `record` calls as you currently use them would have no effect. See the note and last example [here](https://docs.rs/tracing/latest/tracing/span/struct.Span.html#method.record).

**Reply by nickb937** [reply]
It's not the output span that is of concern here, it's logging the same data defined in the span to stdout for collection by some other log-collector like `fluentbit`.  What I desire is having `info!()` calls within a span inherit and output the attributes of the span along with their own attributes.  So `info!("some message")` should output something like `{msg: "some message", "polcy_plan_id": 123, "contract_id": 444, ...}`  

The span fields never appear in the `info!()` log output of subsequent messages, apart from the singular message when the span is created.

**Reply by mladedav** [reply]
That's weird, if you're using `with_current_span`, the output of each longline should be something like this: `{"fields":{"message":"msg","event_field1":"...",...},"span":{"span_field1":"...",...}}`

I'm assuming you're using json output since that's your example, I don't know if the other formatters behave differently.

Anyway if you want something more specific like really moving the span fields together with the event fields, you'd need to use a different subscriber for more control over the output.

**Reply by hawkw** [reply]
It seems like what you want is to flatten all the fields on the current span into the top-level JSON object, rather than emitting a "span" sub-object or a "spans" list. The reason the JSON formatter doesn't do this currently is because an event occurs not only within the current span, but also in that span's parent(s). Using the `with_span_list` option, the whole list of parent spans is included. Span fields are namespaced by the name of the span that defines them, so there may be multiple fields with the same name within the list of all fields defined by parents, so there could be multiple fields with the same name in the trace.

We could probably add an option to include parent span fields in the top level JSON object. However, we'd probably want to do it by adding the span name as a prefix to the field, to resolve potential namespace collisions between fields defined on multiple spans and on the event itself. For example, if we have something like this:

```rust
let _span1 = info_span!("span1", myfield = 1).entered();
let _span2 = info_span!("span2", myfield = 2, something = false).entered();
info!(myfield = 3, "my event");
```
we would want to output JSON like this:
```json
{"message": "my event", "myfield": 3, "span2.myfield": 2, "span2.something": false, "span1.myfield": 1, ...}
```

**Reply by nickb937** [reply]
The handy thing about flattening the namespace is that I can write very simple LogsQL filter like `myfield:"3"` to pull everything associated to `myfield` where value is `3`.   If the span is prefixed onto the field name then it becomes harder.  But I get why you'd prefix the span, since you've no guarantee that the same field name isn't used within the same span for different values, like might happen in a `for..loop`.  Perhaps this just throws up more problems and corner cases than is worth considering.

**Reply by mladedav** [reply]
Sorry, it seems I didn't understand what you were after at all.

There is a PR trying to add what I think you want here #2705

**Reply by nickb937** [reply]
Ah ha, thanks, #2705 is indeed it, https://github.com/mladedav/json-subscriber looks like the path forward.
